### PR TITLE
feat(helm): make change log retrieval more flexible

### PR DESCRIPTION
**Pull request by zanitete**
Make Helm charts changelog retrieval more flexible

## Changes:

* improve the way the Helm chart source url and directory is selected
* use home and source URLs from the latest release instead of the first in the index.yaml array. Reason: Nexus Sonatype Helm repository returns the releases sorted by (probably) creation timestamp.

## Context:

Each Helm chart has its own peculiarities; with this changes we try to cover as many cases as possible to be able to always provide a meaningful change log.

## Documentation

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/helm/index.ts [hunk 1]
```diff
@@ -62,9 +62,10 @@ export class HelmDatasource extends Datasource {
       }
       const result: HelmRepositoryData = {};
       for (const [name, releases] of Object.entries(doc.entries)) {
-        const { sourceUrl, sourceDirectory } = findSourceUrl(releases[0]);
+        const refRelease = releases[releases.length - 1];
```
Comment (viceice) [inline lib/datasource/helm/index.ts:65]: Are you sure the latest release is the last? I'm pretty sure I often saw the opposite. So releases should be sorted with semver.

File: lib/datasource/helm/common.ts [hunk 2]
```diff
@@ -11,30 +11,46 @@ const githubRelease = regEx(
 
 export function findSourceUrl(release: HelmRelease): RepoSource {
   // it's a github release :)
-  let match = githubRelease.exec(release.urls[0]);
-  if (match) {
-    return { sourceUrl: match[1] };
+  const releaseMatch = githubRelease.exec(release.urls[0]);
```
Comment (viceice) [inline lib/datasource/helm/common.ts:14]: Please create tests for all cases, your changes are looking somehow wrong to me.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/datasource/helm/index.ts (hunk 1 above)
Comment (zanitete) [reply lib/datasource/helm/index.ts:65]: you're right. It looks like a Sonatype Nexus specific issue then.

File: lib/datasource/helm/common.ts (hunk 2 above)
Comment (zanitete) [reply lib/datasource/helm/common.ts:14]: Sure, I will. Is there something specific that looks wrong?

File: lib/datasource/helm/common.ts [hunk 3]
```diff
@@ -40,6 +41,19 @@ export function findSourceUrl(release: HelmRelease): RepoSource {
...
 
-  // fallback
-  return { sourceUrl: release.sources[0] };
+  // fallback: if neither home nor sources are a chart repo URL, use githubUrl (if present)
+  const firstSourceMatch = githubUrl.exec(release.sources[0]);
+  if (homeMatch?.groups.url && homeMatch?.groups.path) {
+    return {
+      sourceUrl: homeMatch.groups.url,
+      sourceDirectory: homeMatch.groups.path,
+    };
```
Comment (viceice) [inline lib/datasource/helm/common.ts:50]: When this case will happen?

Comment (zanitete) [reply lib/datasource/helm/common.ts:50]: When the `homeUrl` is a valid Github URL (matches `githubUrl` regexp) but the repo name it's not a valid chart repo (according to the `chartRepo` regexp)

File: lib/datasource/helm/common.ts [hunk 4]
```diff
@@ -16,12 +16,13 @@ export function findSourceUrl(release: HelmRelease): RepoSource {
     return { sourceUrl: releaseMatch[1] };
   }
 
+  let homeMatch;
```
Comment (viceice) [inline lib/datasource/helm/common.ts:19]:
```suggestion
  let homeMatch: { url: string; repo: string } | undefined;
```

File: lib/datasource/helm/common.ts [hunk 5]
```diff
@@ -16,12 +16,13 @@ export function findSourceUrl(release: HelmRelease): RepoSource {
...
   if (release.home) {
-    const githubUrlMatch = githubUrl.exec(release.home);
-    if (githubUrlMatch?.groups && chartRepo.test(githubUrlMatch?.groups.repo)) {
+    homeMatch = githubUrl.exec(release.home);
+    if (homeMatch?.groups && chartRepo.test(homeMatch?.groups.repo)) {
```
Comment (viceice) [inline lib/datasource/helm/common.ts:22]:
```suggestion
    homeMatch = githubUrl.exec(release.home)?.groups;
    if (homeMatch && chartRepo.test(homeMatch.repo)) {
```
 Make other code equivalent.

File: lib/datasource/helm/common.ts [hunk 6]
```diff
@@ -16,12 +16,13 @@ export function findSourceUrl(release: HelmRelease): RepoSource {
...
       return {
-        sourceUrl: githubUrlMatch.groups.url,
-        sourceDirectory: githubUrlMatch.groups.path,
+        sourceUrl: homeMatch.groups.url,
+        sourceDirectory: homeMatch.groups.path,
```
Comment (viceice) [inline lib/datasource/helm/common.ts:25]:
```suggestion
        sourceUrl: homeMatch.url,
        sourceDirectory: homeMatch.path,
```

File: lib/datasource/helm/common.ts [hunk 7]
```diff
@@ -40,6 +41,19 @@ export function findSourceUrl(release: HelmRelease): RepoSource {
...
   }
 
-  // fallback
-  return { sourceUrl: release.sources[0] };
+  // fallback: if neither home nor sources are a chart repo URL, use githubUrl (if present)
+  const firstSourceMatch = githubUrl.exec(release.sources[0]);
+  if (homeMatch?.groups.url && homeMatch?.groups.path) {
+    return {
+      sourceUrl: homeMatch.groups.url,
+      sourceDirectory: homeMatch.groups.path,
```
Comment (viceice) [inline lib/datasource/helm/common.ts:49]:
```suggestion
  if (homeMatch?.url && homeMatch?.path) {
    return {
      sourceUrl: homeMatch.url,
      sourceDirectory: homeMatch.path,
```

**Comment by rarkins** [comment]
Needs lint fix and updating from main

**Pull request closed without merging**